src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+                        }
+                        // the header slows serialization too much
+                        g.setHeader(null);
+                        readsSoFar.add(g);
+                    }
+                    // done reading, ship what we have
+                    if (!readsSoFar.isEmpty()) {
+                        ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                        c.output(ret);
+                    }
+                }
+            }
+
+            @Override
+            public void finishBundle(Context c) throws Exception {
+                if (null!=reader) {
+                    reader.close();
+                }
+                super.finishBundle(c);
+            }
+
+            private boolean accept(SAMRecord r, SimpleInterval region) {
+                if (r.getReadUnmappedFlag() ||
+                    r.getReferenceName() == null || r.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME) ||
+                    r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+                    // unmapped read. We don't want it, *unless* their mate is in the region
+                    return (!r.getMateUnmappedFlag() && r.getMateAlignmentStart() >= region.getStart());
+                } else {
+                    // mapped read.
+                    // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                    return (r.getStart() >= region.getStart());
+                }
+            }
+        };
+    };
+
+    /**
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info,
+     * and fill that and the correct variants into readContext.
+     */
+    public static DoFn<ContextShard, ContextShard> fillContext(final ReferenceDataflowSource refSource, final PCollectionView<SAMSequenceDictionary> refDictView, final GenomicsFactory.OfflineAuth auth) throws IOException {
+        return new DoFnWLog<ContextShard, ContextShard>("fillContext") {
+            private static final long serialVersionUID = 1L;
+            private transient Stopwatch getRef, queryVariants, sending, total;
+            private int elementCount;
+
+            @Override
+            public void startBundle(Context c) throws Exception {
+                super.startBundle(c);
+                getRef = Stopwatch.createUnstarted();
+                queryVariants = Stopwatch.createUnstarted();
+                sending = Stopwatch.createUnstarted();
+                total = Stopwatch.createUnstarted();
+                elementCount=0;
+            }
+
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                total.start();
+                elementCount++;
+                ContextShard shard = c.element();
+                SAMSequenceDictionary refDict = c.sideInput(refDictView);
+                int margin = 500;
```

Comment:
Actually, shouldn't you be using the `BQSR_REFERENCE_WINDOW_FUNCTION` to determine the reference bases for each read instead of this fixed constant? 500 is not always theoretically enough if there are extremely long clipped sequences around a read.